### PR TITLE
ZEPPELIN-1935 Add jceks stored password support for jdbc interpreter

### PR DESCRIPTION
**Pull request by r-kamath**
### What is this PR for?
Add support for jceks stored password instead of using password in clear text. This security enhancement prevents any user from reading clear passwords from interpreter json using shell/py/spark etc. #1315 is the parent PR which fixed similar a issue in Active Directory (shiro.ini config).

### What type of PR is it?
Improvement

### What is the Jira issue?
ZEPPELIN-1935

### How should this be tested?
Create a keystore file using the hadoop credential commandline, for this the hadoop commons should be in the classpath.

`hadoop credential create jdbc.password -provider jceks://file/user/zeppelin/conf/zeppelin.jceks`

Use the jceks file and the key to configure jdbc interpreter. Example interpreter setting:

#### Following are the supported settings for backward compatibility

- Settings with username and jceks
```
default.driver	org.postgresql.Driver
default.jceks.credentialKey	jdbc.password
default.jceks.file	jceks://file/tmp/zeppelin.jceks
default.url	jdbc:postgresql://rkamath-local-1:5432/
default.user	rk-user
```

- Settings with user name and clear text password
```
default.driver	org.postgresql.Driver
default.url	jdbc:postgresql://rkamath-local-1:5432/
default.user	rk-user
default.password	password1
```

### Screenshots (if appropriate)
screen shot 2017-01-10 at 7 02 12 pm

### Questions:
* Does the licenses files need update? no
* Is there breaking changes for older versions? no
* Does this needs documentation? yes

### CODE REVIEW

File: jdbc/src/main/java/org/apache/zeppelin/jdbc/JDBCInterpreter.java [hunk 1]
```diff
@@ -399,6 +409,34 @@ public Connection run() throws Exception {
...
+            provider.getCredentialEntry(properties.getProperty(JDBC_JCEKS_CREDENTIAL_KEY));
+        if (credEntry != null) {
+          return new String(credEntry.getCredential());
+        } else {
+          throw new InterpreterException("Failed to retrieve password from JCEKS from key: "
+              + properties.getProperty(JDBC_JCEKS_CREDENTIAL_KEY));
+        }
+      } catch (Exception e) {
+        logger.error("Failed to retrieve password from JCEKS", e);
+        logger.error("For file: " + properties.getProperty(JDBC_JCEKS_FILE));
```
Comment (felixcheung) [inline jdbc/src/main/java/org/apache/zeppelin/jdbc/JDBCInterpreter.java]: Let's do one logger.error?

**Comment by felixcheung** [comment]
this is important - is there any backward compatibility concerns with changing the password storage?

**Comment by r-kamath** [comment]
@felixcheung JCEKS is used only when the password field is empty. When password and JCEKS properties are empty the connection attempt is made without password.

File: jdbc/src/main/java/org/apache/zeppelin/jdbc/JDBCInterpreter.java (hunk 1 above)
Comment (r-kamath) [inline jdbc/src/main/java/org/apache/zeppelin/jdbc/JDBCInterpreter.java]: 👍

**Comment by prabhjyotsingh** [comment]
@r-kamath IMO this shouldn't be called "backward compatibility" as these are two ways in which password can provided to interpreter.

Tested locally, works as expected, LGTM!

**Comment by felixcheung** [comment]
Got it.

File: jdbc/src/main/java/org/apache/zeppelin/jdbc/JDBCInterpreter.java (hunk 1 above)
Comment (r-kamath) [inline jdbc/src/main/java/org/apache/zeppelin/jdbc/JDBCInterpreter.java]: e62088a

**Comment by prabhjyotsingh** [comment]
Merging this if no more discussion.

**Pull request closed without merging**